File: apps/api/app/repositories/stock_repository.py
from abc import ABC, abstractmethod

from app.schemas.stock import PriceBar, StockDetail, StockOverview, Timeframe


class StockRepository(ABC):
    @abstractmethod
    def list_stocks(self) -> list[StockOverview]:
        raise NotImplementedError

    @abstractmethod
    def get_stock_detail(self, ticker: str) -> StockDetail | None:
        raise NotImplementedError

    @abstractmethod
    def get_price_history(self, ticker: str, timeframe: Timeframe) -> list[PriceBar]:
        raise NotImplementedError

File: apps/web/repositories/stock-repository.ts
import type { PriceBar, StockDetail, StockOverview, Timeframe } from "@/types/stock";

export interface StockRepository {
  listStocks(): Promise<StockOverview[]>;
  getFeaturedStocks(): Promise<StockOverview[]>;
  getStockDetail(ticker: string): Promise<StockDetail | null>;
  getPriceHistory(ticker: string, timeframe: Timeframe): Promise<PriceBar[]>;
}

